src/main/java/appeng/integration/modules/waila/part/P2PStateWailaDataProvider.java

```diff
@@ -0,0 +1,145 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.integration.modules.waila.part;
+
+
+import java.util.List;
+
+import com.google.common.collect.Iterators;
+
+import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import mcp.mobius.waila.api.IWailaConfigHandler;
+import mcp.mobius.waila.api.IWailaDataAccessor;
+
+import appeng.api.parts.IPart;
+import appeng.core.localization.WailaText;
+import appeng.me.GridAccessException;
+import appeng.parts.p2p.PartP2PTunnel;
+
+
+/**
+ * Provides information about a P2P tunnel to WAILA.
+ */
+public final class P2PStateWailaDataProvider extends BasePartWailaDataProvider
+{
+
+	private static final int STATE_UNLINKED = 0;
+	private static final int STATE_OUTPUT = 1;
+	private static final int STATE_INPUT = 2;
+	public static final String TAG_P2P_STATE = "p2p_state";
+
+	/**
+	 * Adds state to the tooltip
+	 *
+	 * @param part part with state
+	 * @param currentToolTip to be added to tooltip
+	 * @param accessor wrapper for various information
+	 * @param config config settings
+	 *
+	 * @return modified tooltip
+	 */
+	@Override
+	public List<String> getWailaBody( final IPart part, final List<String> currentToolTip, final IWailaDataAccessor accessor, final IWailaConfigHandler config )
+	{
+		if( part instanceof PartP2PTunnel )
+		{
+			NBTTagCompound nbtData = accessor.getNBTData();
+			if( nbtData.hasKey( TAG_P2P_STATE ) )
+			{
+				int[] stateArr = nbtData.getIntArray( TAG_P2P_STATE );
+				if( stateArr.length == 2 )
+				{
+					int state = stateArr[0];
+					int outputs = stateArr[1];
+
+					switch( state )
+					{
+						case STATE_UNLINKED:
+							currentToolTip.add( WailaText.P2PUnlinked.getLocal() );
+							break;
+						case STATE_OUTPUT:
+							currentToolTip.add( WailaText.P2POutput.getLocal() );
+							break;
+						case STATE_INPUT:
+							if( outputs <= 1 )
+							{
+								currentToolTip.add( WailaText.P2PInputOneOutput.getLocal() );
+							}
+							else
+							{
+								currentToolTip.add( String.format( WailaText.P2PInputManyOutputs.getLocal(), outputs ) );
+							}
+							break;
+					}
+				}
+			}
+		}
+
+		return currentToolTip;
+	}
+
+	@Override
+	public NBTTagCompound getNBTData( EntityPlayerMP player, IPart part, TileEntity te, NBTTagCompound tag, World world, BlockPos pos )
+	{
+		if( part instanceof PartP2PTunnel )
+		{
+			PartP2PTunnel tunnel = (PartP2PTunnel) part;
+
+			// The default state
+			int state = STATE_UNLINKED;
+			int outputCount = 0;
+
+			if( !tunnel.isOutput() )
+			{
+				try
+				{
+					outputCount = Iterators.size( tunnel.getOutputs().iterator() );
+					if( outputCount > 0 )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)